torchdata/datapipes/iter/util/randomsplitter.py

```diff
@@ -0,0 +1,167 @@
+        >>> train = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='train')
+        >>> list(train)
+        [2, 3, 5, 7, 8]
+        >>> # Be careful to use the same seed as before when specifying `target` to get the correct split.
+        >>> valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='valid')
+        >>> list(valid)
+        [0, 1, 4, 6, 9]
+    """
+
+    def __new__(
+        cls,
+        source_datapipe: IterDataPipe,
+        weights: Dict[T, Union[int, float]],
+        seed,
+        total_length: Optional[int] = None,
+        target: Optional[T] = None,
+    ):
+        if total_length is None:
+            try:
+                # TODO: This is an issue for DataPipes which only have runtime lengths. Revisit to see if this
+                #       is problematic.
+                total_length = len(source_datapipe)
+            except TypeError:
+                raise TypeError(
+                    "RandomSplitter needs `total_length`, but it is unable to infer it from "
+                    f"the `source_datapipe`: {source_datapipe}."
+                )
+        container = _RandomSplitterIterDataPipe(source_datapipe, total_length, weights, seed)
+        if target is None:
+            return [SplitterIterator(container, k) for k in list(weights.keys())]
+        else:
+            if target in weights.keys():
+                return SplitterIterator(container, target)
+            else:
+                raise KeyError(f"`target={target}` does not match any key in `weights`.")
+
+
+class _RandomSplitterIterDataPipe(IterDataPipe):
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe,
+        total_length: int,
+        weights: Dict[T, Union[int, float]],
+        seed,
+    ):
+        self.source_datapipe: IterDataPipe = source_datapipe
+        self.total_length = total_length
+        self._seed = seed
+        self.norm_weights = self.normalize_weights(weights, total_length)
+        self.keys = list(weights.keys())
+        self.key_to_index = {k: i for i, k in enumerate(self.keys)}
+        self.weights = [self.norm_weights[k] for k in self.keys]
+        self._rng = random.Random(self._seed)
+
+    def draw(self):
+        selected_key = self._rng.choices(self.keys, self.weights)[0]
+        self.weights[self.key_to_index[selected_key]] -= 1
+        return selected_key
+
+    @staticmethod
+    def normalize_weights(weights, total_length: int):
+        total_weight = sum(weights.values())
+        return {k: round(float(w) * total_length / total_weight) for k, w in weights.items()}
+
+    def reset(self):
+        self._rng = random.Random(self._seed)
+        self.weights = [self.norm_weights[k] for k in self.keys]
+
+    def set_seed(self, seed):
+        """
+        Update the `seed`. The new `seed` will be used in the next iteration.
+        """
+        self._seed = seed
```

Comment:
Is that possible to unify seed for all `_RandomSplitterIterDataPipe` from the same source?
We might be able to do so by having the same parent-children pattern as `demux`. The parent class has `set_seed` API, which will be invoked by `DataLoader` per iteration. Then, the children DataPipe would take reference the same seed from parent.

With current implementation, there will be multiple `_RandomSplitterIterDataPipe` in the pipeline and technically each of them should have a different seed generated by `DataLoader`.

Reply:
What does this part mean?

> Is that possible to unify seed for all `_RandomSplitterIterDataPipe` from the same source?

Currently, if you `set_seed` for one child, other children are impacted as well.

```python
train, valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0)
train.set_seed(100)
print(train.main_datapipe._seed)  # 100
print(valid.main_datapipe._seed)  # 100
```

Or do you mean when users call `.random_split()` multiple times at different points in the pipeline, we would like all those different calls to share the same seed?

Reply:
> Currently, if you `set_seed` for one child, other children are impacted as well.

Ohhh, Sorry. This is exact what I expect. Just forget my comment above.